File: src/main/java/tmi/recipe/parser/HeatCrafterParser.java
package tmi.recipe.parser;

import arc.struct.Seq;
import mindustry.type.ItemStack;
import mindustry.type.LiquidStack;
import mindustry.world.Block;
import mindustry.world.blocks.production.HeatCrafter;
import mindustry.world.meta.StatUnit;
import tmi.recipe.Recipe;
import tmi.recipe.RecipeType;
import tmi.recipe.types.HeatMark;

public class HeatCrafterParser extends ConsumerParser<HeatCrafter>{
  {excludes.add(GenericCrafterParser.class);}

  @Override
  public boolean isTarget(Block content) {
    return content instanceof HeatCrafter;
  }

  @Override
  public Seq<Recipe> parse(HeatCrafter crafter) {
    Recipe res = new Recipe(RecipeType.factory);
    res.setBlock(getWrap(crafter));
    res.setTime(crafter.craftTime);

    res.addMaterialRaw(HeatMark.INSTANCE, crafter.heatRequirement).setFloatFormat();

    registerCons(res, crafter.consumers);

    if (crafter.outputItems == null) {
      if (crafter.outputItem != null) res.addProduction(getWrap(crafter.outputItem.item), crafter.outputItem.amount);
    }
    else {
      for (ItemStack item : crafter.outputItems) {
        res.addProduction(getWrap(item.item), item.amount);
      }
    }

    if (crafter.outputLiquids == null) {
      if (crafter.outputLiquid != null) res.addProductionPresec(getWrap(crafter.outputLiquid.liquid), crafter.outputLiquid.amount);
    }
    else {
      for (LiquidStack liquid : crafter.outputLiquids) {
        res.addProductionPresec(getWrap(liquid.liquid), liquid.amount);
      }
    }

    return Seq.with(res);
  }
}

File: src/main/kotlin/tmi/recipe/parser/HeatCrafterParser.kt
package tmi.recipe.parser

import arc.struct.Seq
import mindustry.world.Block
import mindustry.world.blocks.production.HeatCrafter
import tmi.recipe.Recipe
import tmi.recipe.RecipeParser
import tmi.recipe.types.RecipeItemType
import tmi.recipe.RecipeType
import tmi.recipe.types.HeatMark

open class HeatCrafterParser : ConsumerParser<HeatCrafter>() {
  override val excludes: Seq<Class<out RecipeParser<*>>> = Seq.with(GenericCrafterParser::class.java)

  override fun isTarget(content: Block): Boolean {
    return content is HeatCrafter
  }

  override fun parse(content: HeatCrafter): Seq<Recipe> {
    val res = Recipe(
      recipeType = RecipeType.factory,
      ownerBlock = content.getWrap(),
      craftTime = content.craftTime,
    )

    res.addMaterial(HeatMark, content.heatRequirement as Number)
      .setType(RecipeItemType.POWER)
      .floatFormat()

    registerCons(res, *content.consumers)

    if (content.outputItems == null) {
      if (content.outputItem != null) res.addProductionInteger(
        content.outputItem.item.getWrap(), content.outputItem.amount
      )
    }
    else {
      for (item in content.outputItems) {
        res.addProductionInteger(item.item.getWrap(), item.amount)
      }
    }

    if (content.outputLiquids == null) {
      if (content.outputLiquid != null) res.addProductionPersec(
        content.outputLiquid.liquid.getWrap(), content.outputLiquid.amount
      )
    }
    else {
      for (liquid in content.outputLiquids) {
        res.addProductionPersec(liquid.liquid.getWrap(), liquid.amount)
      }
    }

    return Seq.with(res)
  }
}
